### ISSUE TITLE
site should prevent closing of tab or navigating away when an editable text box has new user input

### ISSUE THREAD

**Issue by nonamethanks** [danbooru/danbooru#4686]
Since now keyboard shortcuts cannot be disabled anymore, it's become even more aggravating when you're typing something and for any reason the tag box unfocuses and you lose all of your progress. This issue is not just limited to post pages, where just typing "a" or "d" not in the tagbox moves you to another post and erases everything you've done, but also to the upload page, where accidentally closing erases all progress, and in a more limited way to wikis/forums/BUR submission.

Danbooru should simply prevent you from closing the tab with a popup ("Leave site?"), like most modern sites does.
Example from reddit:
![image](https://user-images.githubusercontent.com/12946050/106301592-2a39b080-6258-11eb-81e0-f0d3114e2084.png)

**Comment by nonamethanks** [comment]
I also want to add that it's very easy on a laptop to unfocus the tagbox by accident. The touchpad is quite sensitive on some models and merely passing the wrist over it while typing is enough to click away, and it's distrupted my tagging many times before.

**Comment by luni3359** [comment]
Related issue #4642

**Comment by BrokenEagle** [comment]
I do something similar with New TISAS, and it can be done by simply binding any function at all to a "beforeunload" event to the window object like the following.

```javascript
$(window).on('beforeunload.danbooru', () => "");
```

I added the namespace "danbooru" to the above in case a user wants to unbind only that event.

**Comment by deusexcalamus** [comment]
The site should not try to save the user from themselves. There are times when I'll decide mid-tagging that I don't really want to upload an image midway through tagging, and just ctrl+w the tab. I'd really not have the site asking me 'hurr durr did you really mean to do that'
Yes I did. Now stop asking.

**Comment by nonamethanks** [comment]
Except that the fault is not just on the users, if just pressing a single key is enough to erase all of your progress with no possible way to recover it.

**Comment by bipface** [comment]
@deusexcalamus perhaps it could be a user setting

**Comment by deusexcalamus** [comment]
> 
> 
> @deusexcalamus perhaps it could be a user setting

With evazion's current trend of getting rid of 'niche' user settings, I doubt it.

**Comment by luni3359** [comment]
I agree that the popup would be annoying. That's why I think saving the inputted text would make the most sense; silent for those that don't intend to come back, helpful for those that messed up.

**Comment by luni3359** [comment]
For reference, GitHub "remembers" when you've typed something in the comment textbox, so if you accidentally leave the page you can return and it'll let you continue from where you left.

**Comment by nonamethanks** [comment]
From the discord, a solution that could make everyone happy is to just copy what github does: https://github.com/github/session-resume/
